src/lightgbm/src/main/scala/LightGBMBase.scala

```diff
@@ -73,17 +73,25 @@ trait LightGBMBase[TrainedModel <: Model[TrainedModel]] extends Estimator[Traine
       categoricalSlotIndexesArr, categoricalSlotNamesArr)
     val trainParams = getTrainParams(numWorkers, categoricalIndexes, dataset)
     log.info(s"LightGBM parameters: ${trainParams.toString()}")
-    val networkParams = NetworkParams(getDefaultListenPort, inetAddress, port)
+    val networkParams = NetworkParams(getDefaultListenPort, inetAddress, port, getUseBarrierExecutionMode)
     val validationData =
       if (get(validationIndicatorCol).isDefined && dataset.columns.contains(getValidationIndicatorCol))
         Some(sc.broadcast(df.filter(x => x.getBoolean(x.fieldIndex(getValidationIndicatorCol))).collect()))
       else None
     val preprocessedDF = preprocessData(df)
     val schema = preprocessedDF.schema
-    val lightGBMBooster = preprocessedDF
-      .mapPartitions(TrainUtils.trainLightGBM(networkParams, getLabelCol, getFeaturesCol, get(weightCol),
-        get(initScoreCol), getOptGroupCol, validationData, log, trainParams, numCoresPerExec, schema))(encoder)
-      .reduce((booster1, _) => booster1)
+    val lightGBMBooster =
+      if (getUseBarrierExecutionMode) {
+        preprocessedDF.rdd.barrier().mapPartitions(TrainUtils.trainLightGBM(networkParams, getLabelCol,
+          getFeaturesCol, get(weightCol), get(initScoreCol), getOptGroupCol, validationData, log, trainParams,
```

Comment:
maybe pull this out into a val because its used in both places 

```
val mapPartitionsFunc = TrainUtils.trainLightGBM(networkParams, getLabelCol, getFeaturesCol, get(weightCol),    get(initScoreCol), getOptGroupCol, validationData, log, trainParams, numCoresPerExec, schema)
```

Reply:
good idea, done!